compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -236,137 +142,640 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                            labels.push((missing_span, format!("missing argument of type {}", expected_type)));
+                            if issue_count > 1 {
+                                suggestions.push((missing_span, format!(" {{{}}},", expected_type)));
+                            }
+                        }
+                        Issue::Swap(arg, other) => {
+                            let first_span = provided_args[arg].span;
+                            let second_span = provided_args[other].span;
+                            let first_snippet = source_map.span_to_snippet(first_span).unwrap();
+                            let second_snippet = source_map.span_to_snippet(second_span).unwrap();
+                            let expected_types = (expected_input_tys[arg], expected_input_tys[other]);
+                            let found_types = (
+                                final_arg_types.iter().filter(|(i, _, _)| *i == arg).next().unwrap().1,
+                                final_arg_types.iter().filter(|(i, _, _)| *i == other).next().unwrap().1,
+                            );
+                            labels.push((first_span, format!("expected {}, found {}", expected_types.0, found_types.0)));
+                            suggestions.push((first_span, second_snippet));
+                            labels.push((second_span, format!("expected {}, found {}", expected_types.1, found_types.1)));
+                            suggestions.push((second_span, first_snippet));
+                        }
+                        Issue::Permutation(args) => {
+                            for (src, &arg) in args.iter().enumerate() {
+                                if let Some(dst) = arg {
+                                    let src_span = provided_args[src].span;
+                                    let dst_span = provided_args[dst].span;
+                                    let snippet = source_map.span_to_snippet(src_span).unwrap();
+                                    let expected_type = expected_input_tys[dst];
+                                    let found_type = final_arg_types.iter().filter(|(i, _, _)| *i == dst).next().unwrap().1;
+                                    labels.push((dst_span, format!("expected {}, found {}", expected_type, found_type)));
+                                    suggestions.push((dst_span, snippet));
+                                }
+                            }
+                        }
+                    }
+                }
+
+                // Now construct our error from the various things we've labeled
+                let highlight_sp = MultiSpan::from_spans(labels.iter().map(|s| s.0).collect());
+                let mut err = struct_span_err!(
+                    tcx.sess,
+                    highlight_sp,
+                    E0059, // FIXME: Choose a different code?
+                    "multiple arguments to this function are incorrect"
```

Comment:
Can you only say 'multiple' if there's more than one that's wrong? I think you can tell by looking at `issue_count`.